casadm/cas_lib.c

```diff
@@ -1803,58 +1803,61 @@ int add_core(unsigned int cache_id, unsigned int core_id, const char *core_devic
 	return SUCCESS;
 }
 
-int _check_if_mounted(int cache_id, int core_id)
+bool _device_mounts_detected(int cache_id, int core_id)
 {
 	FILE *mtab;
 	struct mntent *mstruct;
 	char dev_buf[80];
-	int difference = 0, error = 0;
+	int no_match = 0, error = 0, cmplen = 0;
 	if (core_id >= 0) {
 		/* verify if specific core is mounted */
-		snprintf(dev_buf, sizeof(dev_buf), "/dev/cas%d-%d", cache_id, core_id);
+		cmplen = snprintf(dev_buf, sizeof(dev_buf), "/dev/cas%d-%d", cache_id, core_id);
 	} else {
-		/* verify if any core from given cache is mounted */
-		snprintf(dev_buf, sizeof(dev_buf), "/dev/cas%d-", cache_id);
+		/* verify if any core from given cache is mounted
+		   do not compare terminating NULL for cache */
+		cmplen = snprintf(dev_buf, sizeof(dev_buf), "/dev/cas%d-", cache_id) - 1;
 	}
 
 	mtab = setmntent("/etc/mtab", "r");
 	if (!mtab)
 	{
-		cas_printf(LOG_ERR, "Error while accessing /etc/mtab\n");
-		return FAILURE;
+		/* if /etc/mtab not found then the kernel will check for mounts */
+		return false;
 	}
 
 	while ((mstruct = getmntent(mtab)) != NULL) {
-		error = strcmp_s(mstruct->mnt_fsname, PATH_MAX, dev_buf, &difference);
+		error = strcmp_s(mstruct->mnt_fsname, cmplen, dev_buf, &no_match);
 		/* mstruct->mnt_fsname is /dev/... block device path, not a mountpoint */
 		if (error != EOK)
-			return FAILURE;
-		if (!difference) {
-			if (core_id<0) {
-				cas_printf(LOG_ERR,
-					   "Can't stop cache instance %d. Device %s is mounted!\n",
-					   cache_id, mstruct->mnt_fsname);
-			} else {
-				cas_printf(LOG_ERR,
-					   "Can't remove core %d from cache %d."
-					   " Device %s is mounted!\n",
-					   core_id, cache_id, mstruct->mnt_fsname);
+			return false;
+		if (no_match)
+			continue;
+
+		if (core_id < 0) {
+			cas_printf(LOG_ERR,
+				   "Can't stop cache instance %d. Device %s is mounted!\n",
+				   cache_id, mstruct->mnt_fsname);
+		} else {
+			cas_printf(LOG_ERR,
+				   "Can't remove core %d from cache %d."
+				   " Device %s is mounted!\n",
+				   core_id, cache_id, mstruct->mnt_fsname);
 			}
```

Comment:
This function should only check whether cores are mounted. The error messages should be printed by the caller function